### PR TITLE
feat: add image from outside

### LINKED ISSUE #1447: Support dragging and dropping single or multiple images from outside the browser into the editor for adding
This is a very natural way to add local images to the AFFiNE editor. We hope it can be supported based on feedback from the community.

### PR DESCRIPTION
**Pull request by lzlme**
resolve #1447

### CODE REVIEW

**Comment by thorseraq** [comment]
Hi, here's a suggestion: Use `DragIndicator` to indicate the insert position. And insert position should be free, rather than constrained to a paragraph block. To achieve this, `dragenter` event can help to detect when the mouse is dragged from another application to the web page. This way we can leverage `DefaultSelectionManager` to manage the selection state. Then `getBlockEditingStateByCursor()` can be used to find the insertion point and call out `DragHandle`.

**Review by thorseraq — DISMISSED** [review]
Hi, here's a suggestion: Use DragIndicator to indicate the insert position. And insert position should be free, currently it's constrained to a paragraph block.
To achieve this, `dragenter` event can help to detect when the mouse is dragged from another application to the web page. This way we can leverage DefaultSelectionManager to manage the selection state. Then getBlockEditingStateByCursor() can be used to find the insertion point and call out DragHandle.

**Comment by darkskygit** [comment]
> Hi, here's a suggestion: Use DragIndicator to indicate the insert position. And insert position should be free, currently it's constrained to a paragraph block. To achieve this, `dragenter` event can help to detect when the mouse is dragged from another application to the web page. This way we can leverage DefaultSelectionManager to manage the selection state. Then getBlockEditingStateByCursor() can be used to find the insertion point and call out DragHandle.

I add an idea about dragging here: https://github.com/toeverything/blocksuite/issues/1447#issuecomment-1454796934

**Comment by lzlme** [comment]
> > Hi, here's a suggestion: Use DragIndicator to indicate the insert position. And insert position should be free, currently it's constrained to a paragraph block. To achieve this, `dragenter` event can help to detect when the mouse is dragged from another application to the web page. This way we can leverage DefaultSelectionManager to manage the selection state. Then getBlockEditingStateByCursor() can be used to find the insertion point and call out DragHandle.
> 
> I add an idea about dragging here: [#1447 (comment)](https://github.com/toeverything/blocksuite/issues/1447#issuecomment-1454796934)

This looks great, but might be a little difficult for me, I'll try to get it done in the next two days!

**Comment by lzlme** [comment]
@darkskygit Hi, sorry for being late.
I think I finished a general drag scheme and a `registerHandler`, but since I'm not sure if the add image feature is done by blocksuite or affine, I didn't implement this specific feature.

File: packages/editor/src/utils/outside-drag-manager.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import type { BaseBlockModel } from '@blocksuite/store';
+
+import type { EditorContainer } from '../components/editor-container.js';
+
+interface OutsideDragHandler {
+  filter: (files: FileList) => boolean;
+  handler: () => void;
+}
+
+export class OutsideDragManager {
```
Comment (darkskygit) [inline packages/editor/src/utils/outside-drag-manager.ts:17]: cc @lawvs

File: packages/editor/src/utils/outside-drag-manager.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  private _indicator!: DragIndicator;
+  private _disposables = new DisposableGroup();
+  private _cursor = 0;
+  private _getAllowedBlocks: () => BaseBlockModel[] = () => [];
+  private _handlers: Array<OutsideDragHandler> = [];
+
+  constructor(editor: EditorContainer) {
+    setTimeout(() => {
```
Comment (darkskygit) [inline packages/editor/src/utils/outside-drag-manager.ts:25]: may not be best practice, cc @lawvs

**Comment by lzlme** [comment]
> Thank you very much for your contribution. Please let us know when you are ready. We may need to make some adjustments to the code style before merging.

I'm ready. I apologize for the poor code style, and look forward to your guidance.

**Comment by lawvs** [comment]
I apologize for the delay in response. This PR has been blocked by #1447. We will deal with it as soon as the architecture-level design is complete. We will notify you after we are ready. You can rest a little before that.

**Review by Himself65 — COMMENTED** [review]
we have update the blob storage logic

**Comment by doodlewind** [comment]
This is continued in #3123 and has just landed, thanks for starting this @lzlme!

**Pull request closed without merging**